### PR TITLE
per-node configuration overrides

### LINKED ISSUE #22036: CFP: Per-Node configuration overrides
### Background

There is sometimes a need to update or override configuration on one or more nodes. Some possible use-cases include
- Rolling out disruptive changes
- Rolling out changes that need external coordination
- Deploying Cilium in to an existing cluster

#### 1: Rolling out kube-proxy-replacement
Consider a cluster with kube-proxy. The administrator would like to enable Cilium's kube-proxy-replacement feature (KPR). However, to do this, each node must "simultaneously" disable the kube-proxy daemonset as well as enable cilium KPR. Labels on a node naturally lend themselves to this situation

#### 2: Rolling out Cilium to an existing cluster
Imagine deploying Cilium node-by-node in an existing cluster. There are two sets of nodes: migrated and unmigrated. These two sets of nodes might need very different configuration.

#### 3: Platform-managed, administrator overridden
In a cluster where Cilium is maintained by a Platform team, individual cluster owners still may need the ability to tweak certain settings.

### CiliumConfigOverrides CRD
We create a new CRD, `CiliumConfigOverride`. This would allow for overriding arbitrary key-value pairs from the `cilium-config` ConfigMap.

CiliumConfigOverride would have a NodeSelector indicating to which nodes it applied. In the event of conflicting overrides, the lexicograpical ordering of CR names would be used.

Exampe:
```yaml
apiVersion: cilium.io/v2
kind: CiliumConfig
metadata:
  name: run-kube-proxy
spec:
  nodeSelector:
    - matchLabels:
        io.cilium.migration/kube-proxy-replacement: true
  rawConfig:
    kube-proxy-replacement: strict 
```

An empty NodeSelector would apply to all nodes.

### Node annotations
For quick, one-off changes, individual keys could be overridden with annotations on the Node. `io.cilium.config/<KEY>: <VALUE>`

### Configuration resolution

This CR would be read by an InitContainer, which would produce a folder with structure identical to the pre-existing `cilium-config` ConfigMap. Nothing would change for the configuration parsing code in the agent or operator (for now).

By default, the order of precedence would be ConfigMap, then Overrides, then Node annotations. To support case 3, however, the resolution order will be configurable.

There is a more detailed design document at https://docs.google.com/document/d/1_1AsgmY2W2apUUttt0WOoGqlyVEPTK1REokzON-RmR8

### PR DESCRIPTION
**Pull request by squeed**
This adds the ability to tweak configuration on a per-node basis, as designed in #22036. Specifically, there is an InitContainer that implements the configuration resolution logic. It writes a directory that imitates ConfigMaps, so there are no code changes in the agent itself.

This adds a new type, ClilumNodeConfig, that allows for tweaking configuration flags on a fine-grained basis. For example, you could create

```yaml
apiVersion: cilium.io/v2alpha1
kind: CiliumNodeConfig
metadata:
  namespace: kube-system
  name: kube-proxy-replacement
spec:
  selector: 
    matchLabels:
      io.cilium.kube-proxy-replacement: "" 
  rawConfig:
    kube-proxy-replacement: strict
```

then add a NodeAntiAffinity to the kube-proxy daemonset. Then, one by one, label nodes with `io.cilium.kube-proxy-replacement: ""`. This will cause that node to begin using this particular configuration override.

Fixes: #22036

```release-note
Add per-node configuration overrides. There is a new Kubernetes resource type, CiliumNodeConfig, which allows for fine-grained configuration of Nodes based on label selectors.
```

(Note: This also includes a CI workaround for the extra roles; I'll fix that in cilium-cli and remove it shortly).

### CODE REVIEW

**Comment by squeed** [comment]
@varunmar do you want to see if this solves your use-case? Especially the tunable sources?

**Comment by squeed** [comment]
FYI, checks are going to fail until cilium-cli merges and cuts a release.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (kaworu) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: nit: this change is needed because `/tmp/cilium/config-final` is the `build-config.go` default dest right? If correct, how about either a comment or explicitly `--dest` a few lines below?

**Review by kaworu — CHANGES_REQUESTED** [review]
Nice work @squeed !

[This CI complain](https://github.com/cilium/cilium/actions/runs/3505020248/jobs/5871077821) seems legit:

> ciliumnodeconfigs not found in /home/runner/work/cilium/cilium/install/kubernetes/cilium/Chart.yaml.
> Please update the chart to include ciliumnodeconfigs.

Other than that, Helm part of the patch LGTM. `configSources` config feels a bit complex to get right, but admittedly I don't have much context so I'll let other reviewers chim in on the build-config flags.

File: pkg/option/resolver/resolver.go
Comment (thorn3r) [inline pkg/option/resolver/resolver.go]:
```suggestion
// ReadNodeOverrides reads all the CiliumNodeConfig objects and returns a flattened map
```

File: pkg/option/resolver/resolver.go
Comment (thorn3r) [inline pkg/option/resolver/resolver.go]:
```suggestion
			return nil, nil, fmt.Errorf("could not list CiliumNodeConfig objects: %w", err)
```

File: pkg/option/resolver/resolver.go
Comment (thorn3r) [inline pkg/option/resolver/resolver.go]:
```suggestion
			return nil, nil, fmt.Errorf("could not retrieve CiliumNodeConfig %s/%s: %w", namespace, name, err)
```

File: install/kubernetes/cilium/templates/cilium-preflight/clusterrole.yaml
Comment (thorn3r) [inline install/kubernetes/cilium/templates/cilium-preflight/clusterrole.yaml]:
```suggestion
  - ciliumnodeconfigs
```

File: install/kubernetes/cilium/values.yaml
Comment (thorn3r) [inline install/kubernetes/cilium/values.yaml:533]: from a user-perspective, an `any` key like `blockedConfigOverrides: *` might be nice vs. setting an invalid key for `allowedConfigOverrides` to block overrides

**Review by thorn3r — CHANGES_REQUESTED** [review]
awesome feature add, it looks great! there are still some leftover references to `CiliumConfigOverride` in comments and logs it seems. i didn't add a comment for each, but maybe `git grep -i CiliumConfigOverride`

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumnodeconfigs.yaml
Comment (joestringer) [inline pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumnodeconfigs.yaml]: nit: `nodeSelector` would be a bit more consistent with other naming schemes we've used:
```suggestion
              nodeSelector:
```

File: install/kubernetes/cilium/values.yaml
Comment (joestringer) [inline install/kubernetes/cilium/values.yaml]: Do you think there are likely to be any security impacts of these defaults? It feels a bit like we'll have to make some assumptions about the ServiceAccounts/ClusterRoles that k8s application developers typically have and whether this could potentially allow unintentional reconfiguration of cilium-agent configurations on the node. For instance, if node annotations can be freely customized, then that same privilege could be used to disable network policies in Cilium.

File: install/kubernetes/cilium/values.yaml
Comment (joestringer) [inline install/kubernetes/cilium/values.yaml]: Is there an example invalid key that you could suggest? It's not so immediately obvious to me how I would disable this. For that matter, I'd probably just suggest to change `configSources` if the user wanted to completely disable overrides.

File: pkg/option/resolver/resolver_test.go
Comment (joestringer) [inline pkg/option/resolver/resolver_test.go:35]: Do you think it makes sense to have some golden tests or similar for evaluating that this package does the right thing?

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (joestringer) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:711]: nit: Might be nice to add a comment to highlight that this is declared to share the path between the config initContainer and the main container.

File: Documentation/helm-values.rst
Comment (joestringer) [inline Documentation/helm-values.rst]: Curious how this got added, is this automatic with the newer m2r? I didn't notice before. Does it render correctly?

**Review by joestringer — CHANGES_REQUESTED** [review]
Others seemed to be already paying closer attention to the details in pkg/option and cilium/cmd, so I skipped over those. I looked at the rest and from a higher level. Mostly just minor nits, one security-relevant question below.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (squeed) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: You know what? This change isn't needed. I'll get rid of it.

File: install/kubernetes/cilium/values.yaml
Comment (squeed) [reply install/kubernetes/cilium/values.yaml:533]: very good point; I'll add that sugar.

File: install/kubernetes/cilium/values.yaml
Comment (squeed) [reply install/kubernetes/cilium/values.yaml]: I had the same thought around node annotations a few days ago. The Gateway API folks have a good design for a security-sensitive priority-inverting overrides system, which is cool but I balked on implementing.

For now, I'll just disable node annotations by default.

File: install/kubernetes/cilium/values.yaml
Comment (squeed) [reply install/kubernetes/cilium/values.yaml]: Someone suggested '*', but this is even easier.

File: Documentation/helm-values.rst
Comment (squeed) [reply Documentation/helm-values.rst]: I'm not sure; I definitely didn't add this manually. Is there an easy way to preview docs?

File: pkg/option/resolver/resolver_test.go
Comment (squeed) [reply pkg/option/resolver/resolver_test.go:35]: very good point. I'll look in to that.

File: install/kubernetes/cilium/values.yaml
Comment (squeed) [reply install/kubernetes/cilium/values.yaml:533]: I added "allowedConfigOverrides: none" (which isn't actually a code change), and at Joe's suggestion, also suggested disabling overrides above.

File: pkg/option/resolver/resolver_test.go
Comment (squeed) [reply pkg/option/resolver/resolver_test.go:35]: This will take a bit of time to write up, since it needs its own suite. Do you mind if it is in a follow-up PR?

I absolutely think full testing of this is critical, and I won't let it slide :-).

**Comment by squeed** [comment]
OK, review comments addressed, and cilium-cli changes have merged and been released. This should theoretically pass CI now.

File: Documentation/helm-values.rst
Comment (joestringer) [reply Documentation/helm-values.rst]: https://docs.cilium.io/en/latest/contributing/docs/docstest/

File: pkg/option/resolver/resolver_test.go
Comment (joestringer) [reply pkg/option/resolver/resolver_test.go:35]: I'll trust you, sure. :innocent: https://github.com/cilium/cilium/issues/22507

**Comment by squeed** [comment]
Aha, looks like this breaks install-uninstall-install, since there are duplicate RBAC roles.

Filed https://github.com/cilium/cilium-cli/issues/1257 to fix this permanently.

I'll hack over this for now in CI and fix that issue separately.

**Review by kaworu — CHANGES_REQUESTED** [review]
@squeed [Travis-CI is unhappy](https://app.travis-ci.com/github/cilium/cilium/jobs/590054284):

> HINT: to fix this, run 'make -C Documentation update-cmdref'

File: Documentation/helm-values.rst
Comment (squeed) [reply Documentation/helm-values.rst]: Perfect. It doesn't seem to change the rendering, as far as I can see.

**Comment by squeed** [comment]
Filed https://github.com/cilium/cilium/issues/22517 to track removing the CI workaround.

**Comment by squeed** [comment]
Codeql failure seems random: `Installed Go version 1.17 does not match requested Go version 1.19` -- anyone know what this is?

**Comment by squeed** [comment]
CI looks good, except for a CodeQL failure that seems unrelated. If aws-cni and eks passes, this will be good to go.

**Comment by squeed** [comment]
Update: codeql can be ignored according to @aanm.

**Review by thorn3r — APPROVED** [review]
thanks for addressing those comments. LGTM!

**Comment by squeed** [comment]
OK, CI is green, this is ready for any final approvals.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (kaworu) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:363]: is `/tmp` the best path? It seems very generic to me, if it's all the same why not something like `/etc/cilium`?

**Review by kaworu — APPROVED** [review]
Helm changes LGTM

**Review by nebril — APPROVED** [review]
Looking pretty cool!

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumnodeconfigs.yaml
Comment (joestringer) [reply pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumnodeconfigs.yaml]: I think the description got updated but not the field name?

File: install/kubernetes/cilium/README.md
Comment (joestringer) [inline install/kubernetes/cilium/README.md]: Hmm, formatting didn't turn out super great here. Broadly speaking I think that the explanation for this should probably live somewhere dedicated rather than in the configmap and generated to here, because this just ends up in a table column so it's gonna be quite hard to read.

Should we have a new page for this feature from the following location to explain how this works? https://docs.cilium.io/en/latest/configuration/

**Review by joestringer — APPROVED** [review]
Latest changes make sense.

Just a couple of remaining items which you could either address here before we merge or follow up on to tidy up in a fresh PR.

File: install/kubernetes/cilium/README.md
Comment (joestringer) [reply install/kubernetes/cilium/README.md]: Possible other options here: 
- We can leave this as-is, maybe ideally improve the tool that generates this output so that the giant empty spaces are collapsed to 1-2 spaces.
- Move the explanation to docs page, then from the values.yaml and here we remove the explanation (or link to the corresponding docs page)
- Any other ideas?

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumnodeconfigs.yaml
Comment (squeed) [reply pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumnodeconfigs.yaml]: no, I derped the json tag. Fix incoming.

File: install/kubernetes/cilium/README.md
Comment (squeed) [reply install/kubernetes/cilium/README.md]: I definitely intend to write up some pretty extensive documentation about configuration overrides; I'll file a follow-up issue to fix this.

Comment (squeed) [reply install/kubernetes/cilium/README.md]: Cleaned up the formatting a bit, #22576 for more docs.

File: pkg/option/resolver/resolver.go
Comment (joestringer) [inline pkg/option/resolver/resolver.go]: nit: Just FYI, as a convention we typically define the subsystem derived from the directory path, so in this case `option-resolver`.

File: pkg/option/resolver/resolver.go
Comment (joestringer) [inline pkg/option/resolver/resolver.go]: nit: We typically try to avoid formatting variations on logging statements, because those break structured logging and log aggregation systems since the strings are dynamic. In this case it doesn't matter quite as much since the code will be used from a separate commandline utility, but typically we would instead define this explicitly such as:
```suggestion
	log.WithFields(logrus.Fields{
	    logfields.Source: source,
	}).Info("Reading option configuration")
```

(Same for all Infof(), Errorf(), etc log statements in this file)

**Review by joestringer — APPROVED** [review]
I did one last review pass on this. There's a couple of logging aspects that are written in a style we typically try to avoid, but given that the library is only used by this separate CLI it's not such a big deal. I think the next step from a project perspective would be to formalize those linters so that we are consistent with applying those principles, rather than having folks like me point out the common style during review like this. We can already see how easy it is to miss this sort of feedback.

**Comment by joestringer** [comment]
Hmm, Travis is failing so I cannot merge this PR. Please fix the remaining CI issues.

**Comment by christarazi** [comment]
Glancing through this PR, one thing that popped out at me is that the CiliumNodeConfig CRD's key name format is different than the configmap key format. Is there a reason why that is? Does that mean there's a translation layer somewhere to do this and if so, is it tested?

**Comment by squeed** [comment]
> Glancing through this PR, one thing that popped out at me is that the CiliumNodeConfig CRD's key name format is different than the configmap key format.

That shouldn't be the case -- the code treats all configuration keys the same...

**Comment by squeed** [comment]
> Hmm, Travis is failing so I cannot merge this PR. Please fix the remaining CI issues.

argh! Fixed. I always forget one of the million `make`s that this PR needs.

File: pkg/option/resolver/resolver.go
Comment (squeed) [reply pkg/option/resolver/resolver.go]: OK, I added structured logging. Good catch.

File: pkg/option/resolver/resolver.go
Comment (squeed) [reply pkg/option/resolver/resolver.go]: fixed.

**Comment by christarazi** [comment]
> > Glancing through this PR, one thing that popped out at me is that the CiliumNodeConfig CRD's key name format is different than the configmap key format.
> 
> That shouldn't be the case -- the code treats all configuration keys the same...

In the CiliumNodeConfig CRD, it uses `kubeProxyReplacement`. In the Cilium configmap, it's `kube-proxy-replacement`.

**Comment by squeed** [comment]
Ooh, that’s just an error in the example in the PR description!

**Comment by christarazi** [comment]
Ok great, I've fixed the description so that it's not confusing to future readers.

File: pkg/option/resolver/resolver.go
Comment (joestringer) [reply pkg/option/resolver/resolver.go]: There's still some cases where there's now both structured logging and formatting versions of the log calls. I think the right way forward is we just set up a linter for this though, then we can clean up any remaining cases in the codebase at that time.

**Comment by pchaigno** [comment]
I'd bump this to `release-note/major`. Minor includes a lot of stuff, like deprecated flag removals, new error messages, slight changes in behavior, etc. I think this would benefit from being more visible in release notes.

**Pull request merged**
